util/grpc/src/auth/token_authenticator.rs

```diff
@@ -0,0 +1,217 @@
+    /// tokens to be cryptographically-verified by the authenticator.
+    shared_secret: [u8; 32],
+
+    /// The maximum duration a token is valid for.
+    max_token_lifetime: Duration,
+}
+
+impl Drop for TokenAuthenticator {
+    fn drop(&mut self) {
+        self.shared_secret.zeroize();
+    }
+}
+
+impl Authenticator for TokenAuthenticator {
+    fn authenticate(
+        &self,
+        maybe_credentials: Option<BasicCredentials>,
+    ) -> Result<String, AuthenticatorError> {
+        let credentials = maybe_credentials.ok_or(AuthenticatorError::Unauthenticated)?;
+        let mut parts = credentials.password.split(':');
+        let username = parts
+            .next()
+            .ok_or(AuthenticatorError::InvalidAuthorizationToken)?;
+        let timestamp = parts
+            .next()
+            .ok_or(AuthenticatorError::InvalidAuthorizationToken)?;
+        let signature = parts
+            .next()
+            .ok_or(AuthenticatorError::InvalidAuthorizationToken)?;
+        if parts.next().is_some() {
+            return Err(AuthenticatorError::InvalidAuthorizationToken);
+        }
+        if username != credentials.username {
+            return Err(AuthenticatorError::InvalidAuthorizationToken);
+        }
+        if !self.is_valid_time(timestamp, SystemTime::now())? {
+            return Err(AuthenticatorError::ExpiredAuthorizationToken);
+        }
+        if !self.is_valid_signature(&format!("{}:{}", username, timestamp), signature)? {
+            return Err(AuthenticatorError::InvalidAuthorizationToken);
+        }
+        Ok(credentials.username)
+    }
+}
+
+impl TokenAuthenticator {
+    pub fn new(shared_secret: [u8; 32], max_token_lifetime: Duration) -> Self {
+        Self {
+            shared_secret,
+            max_token_lifetime,
+        }
+    }
+
+    fn is_valid_time(&self, timestamp: &str, now: SystemTime) -> Result<bool, AuthenticatorError> {
+        let token_time: Duration = Duration::from_secs(
+            timestamp
+                .parse()
+                .map_err(|_| AuthenticatorError::InvalidAuthorizationToken)?,
+        );
+        let our_time: Duration = now
+            .duration_since(SystemTime::UNIX_EPOCH)
+            .map_err(|_| AuthenticatorError::ExpiredAuthorizationToken)?;
+        let distance: Duration = our_time
+            .checked_sub(token_time)
+            .unwrap_or_else(|| token_time - our_time);
+        Ok(distance < self.max_token_lifetime)
+    }
+
+    fn is_valid_signature(&self, data: &str, signature: &str) -> Result<bool, AuthenticatorError> {
+        let mut mac = Hmac::<sha2::Sha256>::new_varkey(&self.shared_secret)
+            .map_err(|_| AuthenticatorError::Other("Invalid HMAC key".to_owned()))?;
+        mac.update(data.as_bytes());
+        let our_signature = mac.finalize().into_bytes();
+
+        let their_suffix: Vec<u8> =
+            hex::decode(signature).map_err(|_| AuthenticatorError::InvalidAuthorizationToken)?;
```

Comment:
This can probably go at the top of this function, as signal is doing it. I don't think it precludes timing attacks against `new_varkey()`, if there are any, but it probably won't hurt to validate user input before accessing secrets... ;)